File: apps/ios/PicooCamera/ManualEndpointInput.swift
import SwiftUI

// REQ-PICOO-UI-003 / AC-M-MANUAL-01
nonisolated struct ManualEndpointDraft: Equatable, Sendable {
    static let defaultPort = "4433"

    var octets: [String]
    var port: String

    init(text: String) {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let separator = trimmed.lastIndex(of: ":")
        let hostText = separator.map { String(trimmed[..<$0]) } ?? trimmed
        let portText = separator.map { String(trimmed[trimmed.index(after: $0)...]) } ?? Self.defaultPort
        let parsedOctets = hostText.split(separator: ".", omittingEmptySubsequences: false)
        octets = (0..<4).map { index in
            guard parsedOctets.indices.contains(index) else { return "" }
            return String(parsedOctets[index]).decimalDigits(maxLength: 3)
        }
        port = separator == nil ? Self.defaultPort : portText.decimalDigits(maxLength: 5)
    }

    init?(pastedText: String) {
        let colonCount = pastedText.filter { $0 == ":" }.count
        let hostText = pastedText.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false).first ?? ""
        guard colonCount <= 1,
              hostText.split(separator: ".", omittingEmptySubsequences: false).count == 4
        else { return nil }
        self.init(text: pastedText)
        guard octets.allSatisfy({ !$0.isEmpty }) else { return nil }
    }

    var text: String { "\(octets.joined(separator: ".")):\(port)" }

    mutating func updateOctet(_ index: Int, value: String) {
        octets[index] = value.decimalDigits(maxLength: 3)
    }

    mutating func updatePort(_ value: String) {
        port = value.decimalDigits(maxLength: 5)
    }

    static func shouldAdvance(octet: String) -> Bool {
        let digits = octet.decimalDigits(maxLength: 3)
        return digits.count == 3 || (digits.count == 2 && (Int(digits) ?? 0) > 25)
    }
}

nonisolated private extension String {
    func decimalDigits(maxLength: Int) -> String {
        String(filter { "0123456789".contains($0) }.prefix(maxLength))
    }
}

struct IPv4EndpointInput: View {
    @Binding var endpoint: String
    @FocusState private var focusedField: Field?

    private enum Field: Hashable {
        case octet(Int)
        case port
    }

    var body: some View {
        let draft = ManualEndpointDraft(text: endpoint)
        VStack(alignment: .leading, spacing: PicooSpace.xs) {
            Text("IP 地址")
                .font(.caption)
                .foregroundStyle(PicooColor.contentMuted)

            HStack(spacing: PicooSpace.xs) {
                ForEach(draft.octets.indices, id: \.self) { index in
                    endpointField(
                        value: draft.octets[index],
                        placeholder: index == 0 ? "192" : index == 1 ? "168" : "1",
                        label: "IP 地址第 \(index + 1) 段",
                        field: .octet(index)
                    ) { value in
                        updateOctet(index, value: value)
                    }
                    if index != draft.octets.indices.last {
                        Text(".").font(.headline)
                    }
                }
                Text(":").font(.headline)
                endpointField(
                    value: draft.port,
                    placeholder: ManualEndpointDraft.defaultPort,
                    label: "端口",
                    field: .port
                ) { value in
                    var updated = ManualEndpointDraft(text: endpoint)
                    updated.updatePort(value)
                    endpoint = updated.text
                }
                .frame(maxWidth: 76)
            }
        }
        .toolbar {
            ToolbarItemGroup(placement: .keyboard) {
                Spacer()
                Button(focusedField == .port ? "完成" : "下一项") {
                    advanceFocus()
                }
            }
        }
        .onAppear { focusedField = .octet(0) }
    }

    private func endpointField(
        value: String,
        placeholder: String,
        label: String,
        field: Field,
        onChange: @escaping (String) -> Void
    ) -> some View {
        TextField(
            "",
            text: Binding(get: { value }, set: { newValue in onChange(newValue) }),
            prompt: Text(placeholder)
        )
            .font(.body.monospacedDigit())
            .multilineTextAlignment(.center)
            .keyboardType(.numberPad)
            .focused($focusedField, equals: field)
            .textFieldStyle(.roundedBorder)
            .frame(minWidth: 0, maxWidth: .infinity, minHeight: PicooIconSize.touchTarget)
            .accessibilityLabel(label)
    }

    private func updateOctet(_ index: Int, value: String) {
        if let pasted = ManualEndpointDraft(pastedText: value) {
            endpoint = pasted.text
            focusedField = .port
            return
        }

        var updated = ManualEndpointDraft(text: endpoint)
        updated.updateOctet(index, value: value)
        endpoint = updated.text
        if ManualEndpointDraft.shouldAdvance(octet: value) {
            focusedField = index == 3 ? .port : .octet(index + 1)
        }
    }

    private func advanceFocus() {
        switch focusedField {
        case let .octet(index):
            focusedField = index == 3 ? .port : .octet(index + 1)
        case .port:
            focusedField = nil
        case nil:
            focusedField = .octet(0)
        }
    }
}

File: apps/android/app/src/main/kotlin/com/picoo/camera/ui/ManualEndpointInput.kt
package com.picoo.camera.ui

/** REQ-PICOO-UI-003 / AC-M-MANUAL-01 — editable IPv4 and port field state. */
internal data class ManualEndpointDraft(
    val octets: List<String>,
    val port: String,
) {
    init {
        require(octets.size == IPV4_OCTET_COUNT)
    }

    fun updateOctet(index: Int, value: String): ManualEndpointDraft = copy(
        octets = octets.toMutableList().also { it[index] = value.digits(MAX_OCTET_DIGITS) },
    )

    fun updatePort(value: String): ManualEndpointDraft = copy(port = value.digits(MAX_PORT_DIGITS))

    fun asText(): String = "${octets.joinToString(".")}:$port"

    fun validatedEndpoint(): ManualEndpoint? {
        val values = octets.map { octet ->
            octet.toIntOrNull()?.takeIf { it in 0..255 } ?: return null
        }
        val parsedPort = port.toIntOrNull()?.takeIf { it in 1..65535 } ?: return null
        return ManualEndpoint(values.joinToString("."), parsedPort)
    }

    companion object {
        private const val IPV4_OCTET_COUNT = 4
        private const val MAX_OCTET_DIGITS = 3
        private const val MAX_PORT_DIGITS = 5
        private val DEFAULT_PREFIX = listOf("192", "168")
        const val DEFAULT_PORT = "4433"

        fun from(text: String): ManualEndpointDraft {
            val trimmed = text.trim()
            val separator = trimmed.lastIndexOf(':')
            val hostText = if (separator >= 0) trimmed.substring(0, separator) else trimmed
            val portText = if (separator >= 0) trimmed.substring(separator + 1) else DEFAULT_PORT
            val parsedOctets = hostText.split('.').take(IPV4_OCTET_COUNT)
            val octets = List(IPV4_OCTET_COUNT) { index ->
                parsedOctets.getOrNull(index).orEmpty().digits(MAX_OCTET_DIGITS).ifEmpty {
                    if (hostText.isEmpty()) DEFAULT_PREFIX.getOrNull(index).orEmpty() else ""
                }
            }
            return ManualEndpointDraft(
                octets = octets,
                port = if (separator >= 0) {
                    portText.digits(MAX_PORT_DIGITS)
                } else {
                    DEFAULT_PORT
                },
            )
        }

        fun fromPastedText(text: String): ManualEndpointDraft? {
            val trimmed = text.trim()
            if (trimmed.count { it == ':' } > 1) return null
            val hostText = trimmed.substringBeforeLast(':', trimmed)
            if (hostText.split('.').size != IPV4_OCTET_COUNT) return null
            val draft = from(trimmed)
            return draft.takeIf { it.octets.all(String::isNotEmpty) }
        }

        fun shouldAdvanceOctet(value: String): Boolean {
            val digits = value.digits(MAX_OCTET_DIGITS)
            return digits.length == MAX_OCTET_DIGITS ||
                (digits.length == 2 && (digits.toIntOrNull() ?: 0) > 25)
        }

        private fun String.digits(maxLength: Int): String = filter { it in '0'..'9' }.take(maxLength)
    }
}

internal data class ManualEndpoint(val host: String, val port: Int)
